text/0000-inherent-trait-impl.md

```diff
@@ -0,0 +1,152 @@
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provides a mechanism to declare "inherent traits" for a type defined in the same crate. Methods on these traits are callable on instances of the specified type without needing to import the trait.
+
+# Motivation
+[motivation]: #motivation
+
+There are two similar cases where this is valuable:
+
+- Frequently used traits.
+
+  Sometimes getting the right abstractions require breaking up a type's implementation into many traits, with only a few methods per
+  trait. Every use of such a type results in a large number of imports to ensure the correct traits are in scope. If such a type is used
+  frequently, then this burden quickly becomes a pain point for users of the API,
+  especially if users do not care about writing generic code over traits.
+
+- Mapping object-oriented APIs.
+
+  When mapping these APIs to rust, base classes are usually mapped to traits: methods on those base classes will need to be callable on any
+  derived type. This is sub-optimal because to use a class method a user must now know which class in the hierachy defined that
+  method, so that they can import and use the corresponding trait. This knowledge is not required when using the same API from an
+  object-oriented language.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The feature is implemented using a new attribute which can be applied to trait
+`impl` blocks:
+
+```rust
+pub trait Bar {
+    fn bar(&self);
+}
+
+pub trait ExtBar: Bar {
+    fn ext_bar(&self);
+}
+
+impl<T: Bar> ExtBar for T {
+    fn ext_bar(&self) { }
+}
+
+pub struct Foo;
+
+#[inherent]
+impl Bar for Foo {
+    fn bar(&self) { println!("foo::bar"); }
+}
+
+// works thanks to specialization
+#[inherent]
+impl ExtBar for Foo { }
+
+impl Foo {
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+The methods `bar` and `ext_bar` are now callable on any instance of `Foo`,
+regardless of whether the `Bar` and `ExtBar` traits are currently in scope,
+or even whether the `Bar` trait is publically visible. In other words if `Bar`
+and `ExtBar` defined in one crate and `Foo` and another, user of `Foo` will be
+able to explicitly depend only on the crate which defines `Foo` and still use
+the inherent traits methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `inherent` attribute in the above example makes the `impl` block equivalent to:
+
+```rust
+impl Foo {
+    #[inline]
+    pub fn bar(&self) { <Self as Bar>::bar(self); }
+
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+Any questions regarding coherence, visibility or syntax can be resolved by
+comparing against this expansion, although the feature need not be implemented
+as an expansion within the compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Increased complexity of the language.
+- Hides use of traits from users.
```

Comment:
If this works the way I think it does, this actually changes our trait stability story, and that's a _major_ drawback.

Currently, adding a defaulted method to a trait is not a major breaking change, at worst it will cause ambiguity errors, so it's [considered minor](http://rust-lang.github.io/rfcs/1105-api-evolution.html#minor-change-adding-a-defaulted-item).

Now, a dependency can add a method to a trait you `#[inherent]` impl, which can clash with a method of your own, causing your build to fail in a way that requires you to change your API to fix. We're largely okay with builds failing due to new ambiguities (clashing method names across traits, adding something to a module that's glob imported, etc) and such things are categorized as [minor](http://rust-lang.github.io/rfcs/1105-api-evolution.html#motivation), which basically means it's fine to do as long as the fallout isn't too much. With this RFC, adding a defaulted method has the potential to break a library user in a way that requires them to rename a method in their public API, causing a breaking change for them.

This should be explored and addressed in this RFC, and as a bare minimum should be called out in this section.

(One "fix" is to only allow local traits)

Reply:
Another fix is to list out the method names that are inherent, although that becomes quite a burden for e.g. iterator methods.

Reply:
I'm unsure if the existing situation should be classified as so minor because you interact with many types almost exclusively through their traits.  As an example, there are definitely traits with a `new()` method that defies the convention of being inherent `new()`, so those traits adding `new()` created exactly the same breakage you describe here.

As a fix, I'd suggest `#[inherent]` being a method attribute for items inside an `impl`, so
```
impl TraitWIthBadMethodNames for MyType {
    fn new() -> Self { .. }  // Not inherent

    #[inherent]
    fn new_plus() -> Self;  // Inherent but default body used

    #[inherent]
    fn foo() { .. }  // Inherent with body supplied here
}
```
And `#[inherent]` applied to the `impl` is equivalent to it being applied to all methods and associated types and constants.

Reply:
> As an example, there are definitely traits with a new() method that defies the convention of being inherent new(), so those traits adding new() created exactly the same breakage you describe here.

Yes, but this breakage is easily fixed by using UFCS. This is not true for breakage caused in the world of `#[inherent]`.

The "minor" terminology comes from the API evolution RFC, such changes _may_ cause crates to stop compiling, however:

 - This can be fixed with a trivial change local to the crate
 - The fix does not break upstream crates

Without this notion of "minor" being allowed, crates wouldn't be able to add _anything_ new (types, traits, functions, or methods) without it being considered a breaking change.

Furthermore, the API evolution RFC mentions in the trait method case that you should check to ensure the fallout isn't too great; which is where your `new()` example falls short: a trait adding a `new()` method would probably have lots of fallout.

In other words, when I use the term "minor" here, I'm using a precisely defined term from another RFC. Whether or not it is actually "minor" is irrelevant, I'm talking about what we do and don't consider breaking.

Reply:
Will it be possible to implement shadowing of inherent trait methods by true inherent methods? Compiler will warn on clashing inherent names, but will use true inherent method by default and for trait method you will have to use explicit `Trait::foo(value)`. This way we will avoid code breakage on "minor" upstream changes.

Though I think in practice such collisions should be extremely rare.

Reply:
> Will it be possible to implement shadowing of inherent trait methods by true inherent methods?

I think that would be the ideal fix here: an inherent trait's methods get shadowed by the type's own methods. They could produce a warning lint when compiling the crate itself, so that people notice, and then cap-lints will suppress that when compiling a dependency.